infrastructure/comms/src/peer_manager/node_id.rs

```diff
@@ -0,0 +1,243 @@
+        let mut indices: Vec<usize> = Vec::with_capacity(node_ids.len());
+        let mut dists: Vec<NodeDistance> = Vec::with_capacity(node_ids.len());
+        for i in 0..node_ids.len() {
+            indices.push(i);
+            dists.push(self.distance(&node_ids[i]));
+        }
+        // Perform partial sort of elements only up to K elements
+        let mut nearest_node_indices: Vec<usize> = Vec::with_capacity(k);
+        for i in 0..k {
+            for j in i + 1..node_ids.len() {
+                if dists[i] > dists[j] {
+                    dists.swap(i, j);
+                    indices.swap(i, j);
+                }
+            }
+            nearest_node_indices.push(indices[i]);
+        }
+        Ok(nearest_node_indices)
+    }
+
+    /// Find and return the node ids of the K nearest neighbours from the provided node id list
+    pub fn closest(&self, node_ids: &Vec<NodeId>, k: usize) -> Result<Vec<NodeId>, NodeIdError> {
+        match self.closest_indices(&node_ids, k) {
+            Ok(nearest_node_indices) => {
+                let mut nearest_node_ids: Vec<NodeId> = Vec::with_capacity(nearest_node_indices.len());
+                for i in 0..nearest_node_indices.len() {
+                    nearest_node_ids.push(node_ids[nearest_node_indices[i]].clone());
+                }
+                Ok(nearest_node_ids)
+            },
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl PartialEq for NodeId {
+    fn eq(&self, nid: &NodeId) -> bool {
+        (self.0 == nid.0)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use sha2::Sha256;
+    use tari_crypto::{
+        keys::SecretKey,
+        ristretto::{RistrettoPublicKey, RistrettoSecretKey},
+    };
+    use tari_utilities::byte_array::ByteArray;
+
+    #[test]
+    fn test_from_public_key() {
+        let mut rng = rand::OsRng::new().unwrap();
+        let sk = RistrettoSecretKey::random(&mut rng);
+        let pk = RistrettoPublicKey::from_secret_key(&sk);
+        let node_id = NodeId::from_public_key::<RistrettoPublicKey, Sha256>(&pk);
+        assert_ne!(node_id.0.to_vec(), NodeId::new().0.to_vec());
+        // Ensure node id is different to original public key
+        let mut pk_array: [u8; 32] = [0; 32];
+        pk_array.copy_from_slice(&pk.as_bytes());
+        assert_ne!(node_id.0.to_vec(), pk_array.to_vec());
+    }
+
+    #[test]
+    fn test_distance_and_ordering() {
+        let mut rng = rand::OsRng::new().unwrap();
+        let sk = RistrettoSecretKey::random(&mut rng);
+        let pk = RistrettoPublicKey::from_secret_key(&sk);
+        let node_id1 = NodeId::from_public_key::<RistrettoPublicKey, Sha256>(&pk);
+        let mut node_id2 = node_id1.clone();
+        node_id2.0[16] += 1;
+        let mut node_id3 = node_id1.clone();
+        node_id3.0[24] -= 1;
+        assert!(node_id1.0 < node_id2.0);
+        assert!(node_id1.0 > node_id3.0);
+        assert!(node_id1.distance(&node_id2) > node_id1.distance(&node_id3));
```

Comment:
Perhaps, in addition to this, test that the distance function gives the correct XOR distance value

Reply:
Change test so actual distances are also used.